File: modules/expo-battery/ios/ExpoBatteryModule.swift
import ExpoModulesCore
import ExpoUI

public class ExpoBatteryModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBattery")

    Events("onChange")

    Constant("PI") {
      Double.pi
    }

    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { (value: String) in
      self.sendEvent("onChange", [
        "value": value
      ])
    }

    View(ExpoBatteryView.self) {
      Events("onTap")
    }

    Class(ExpoBatteryModuleSharedObject.self) {
      Constructor { () -> ExpoBatteryModuleSharedObject in
        return ExpoBatteryModuleSharedObject()
      }

      Property("count") { (ref: ExpoBatteryModuleSharedObject) -> Int in
        return ref.count
      }
      .set { (ref: ExpoBatteryModuleSharedObject, count: Int) in
        ref.count = count
      }
    }

    ExpoUIView(ExpoBatterySwiftUIView.self)

    OnCreate {
      ViewModifierRegistry.register("expoBatterySwiftUIModifier") { params, appContext, _ in
        return try ExpoBatterySwiftUIModifier(from: params, appContext: appContext)
      }
    }

    OnDestroy {
      ViewModifierRegistry.unregister("expoBatterySwiftUIModifier")
    }
  }
}

File: modules/expo-battery/android/src/main/java/expo/modules/battery/ExpoBatteryModule.kt
package expo.modules.battery

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.ui.ExpoUIView
import expo.modules.kotlin.records.recordFromMap
import expo.modules.ui.ModifierRegistry

class ExpoBatteryModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBattery")

    Events("onChange")

    Constant("PI") {
      Math.PI
    }

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("setValueAsync") { value: String ->
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }

    View(ExpoBatteryView::class) {
      // Defines an event that the view can send to JavaScript.
      Events("onTap")
    }

    Class(ExpoBatteryModuleSharedObject::class) {
      Constructor {
        val instance = ExpoBatteryModuleSharedObject(appContext)
        return@Constructor instance
      }

      Property("count")
        .get { ref: ExpoBatteryModuleSharedObject ->
          ref.count
        }
        .set { ref: ExpoBatteryModuleSharedObject, count: Int ->
          ref.count = count
        }
    }

    ExpoUIView<ExpoBatteryComposeViewProps>("ExpoBatteryComposeView") {
      Content { props ->
        ExpoBatteryComposeViewContent(props)
      }
    }

    OnCreate {
      ModifierRegistry.register("expoBatteryComposeModifier") { params, _, _, _ ->
        recordFromMap<ExpoBatteryComposeModifierParams>(params).toModifier()
      }
    }
  }
}
